root/static/scripts/edit/URLCleanup.js

```diff
@@ -4826,6 +4826,35 @@ const CLEANUPS: CleanupEntries = {
       return url;
     },
   },
+  'tsutaya': {
+    match: [new RegExp('^(https?://)?shop\\.tsutaya\\.co\\.jp/', 'i')],
+    restrict: [LINK_TYPES.mailorder],
+    clean: function (url) {
+      url = url.replace(/^(?:https?:\/\/)?shop\.tsutaya\.co\.jp\/cd\/product\/(\d+)(?:\/.*)?$/, 'https://shop.tsutaya.co.jp/cd/product/$1/');
+      url = url.replace(/^(?:https?:\/\/)?shop\.tsutaya\.co\.jp\/dir_result\.html\?searchType\=3\&artistCD\=(\d+)\&artistName\=(.*)?$/, 'https://shop.tsutaya.co.jp/dir_result.html?searchType=3&artistCd=$1');
+      return url;
+    },
+    validate: function (url, id) {
+      const m = /^https:\/\/shop\.tsutaya\.co.jp\/(dir_result|cd)\.*$/.exec(url);
+      if (m) {
+        const suffix = m[1];
```

Comment:
There is a buggy backslash `\` in front of the special character dot `.` at the end of the matching pattern. That is the only reason the tests don’t pass successfully.

Note also that it matches only the most distinctive part of valid URLs, but the pattern could be as complete as patterns in the `clean` function are:

```suggestion
      const m = /^https:\/\/shop\.tsutaya\.co.jp\/(?:(cd)\/product\/\d+\/|(dir_result)\.html\?searchType=3\&artistCd=\d+)$/.exec(url);
      if (m) {
        const suffix = m[1] || m[2];
```